src/qt3utils/applications/controllers/random_data_generator.py

```diff
@@ -125,28 +125,186 @@ class QT3ScanRandomDataController(QT3ScopeRandomDataController):
     """
+
     @property
     def clock_rate(self) -> float:
         return self.data_generator.clock_rate
 
     def sample_counts(self, num_batches: int) -> np.ndarray:
         return self.data_generator.sample_counts(num_batches)
 
-    def sample_count_rate(self, data_counts: np.ndarray) -> np.ndarray:
+    def sample_count_rate(self, data_counts: np.ndarray) -> np.floating:
         return self.data_generator.sample_count_rate(data_counts)
 
+
+class QT3ScanRandomSpectrometerDataController:
+    """
+    Implements qt3utils.applications.qt3scan.interface.QT3ScanSpectrometerDAQControllerInterface
+    """
+    def __init__(self, logger_level: int):
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel(logger_level)
+
+        self.spectrometer = self.RandomSpectometer()
+
+        self.last_config_dict = {}
+
+        self.last_measured_spectrum = None
+        self.last_wavelength_array = None
+
+    class RandomSpectometer:
+        def __init__(self):
+            self.exposure_time = 500 # milliseconds
+            self.experiment_name = ""
+            self.num_wavelength_bins = 250
+            self.wave_start = 600
+            self.wave_end = 850
+            self.num_frames = 1
+            self.center_wavelength = 700  # nm
+            self.temperature_sensor_setpoint = -70  # grad
+
+            self.nv_probability = 0.01
+            self.background_counts = int(1e5)
+            self.nv_brightness = int(1e6)
+
+        def acquire_step_and_glue(self) -> Tuple[np.ndarray, np.ndarray]:
+            wavelengths = np.linspace(self.wave_start, self.wave_end, self.num_wavelength_bins)
+            if np.random.random() > self.nv_probability:
+                spectrum = self.background_counts * np.random.random(self.num_wavelength_bins) / self.num_wavelength_bins
+            else:
+                redux = 10
+                num_samples = int(self.nv_brightness / redux) # a little hack to make the sampling faster
+                sample_sideband = np.random.normal(690, 40, size=99 * num_samples // 100)
+                hist_sideband, _ = np.histogram(sample_sideband, bins=range(self.wave_start, self.wave_end + 1))
+                zpl_sample = np.random.normal(637, 2, size=1 * num_samples // 100)
+                zpl, _ = np.histogram(zpl_sample, bins=range(self.wave_start, self.wave_end + 1))
+                spectrum = (zpl + hist_sideband) * redux
+            return spectrum, wavelengths
+
     @property
-    def num_data_samples_per_batch(self) -> int:
-        return self.data_generator.num_data_samples_per_batch
+    def clock_rate(self) -> float:
+        try:
+            _t = self.spectrometer.exposure_time / 1000.0  #Convert from milliseconds to seconds.
+        except Exception as e:
+            self.logger.error(e)
+            _t = 2  #TODO: better default behavior. Should this be -1? 1? or should Spectrometer be changed.
```

Comment:
would np.nan work better?

Reply:
Actually... I think I can remove this whole dumb thing. I copied this from the PrincetonSpectrometerDAQController. (But there's a good reason why it's done this way in that class.)